### DISCUSSION TITLE
router.events.on not firing

### DISCUSSION THREAD

**Discussion by jamiviz** [vercel/next.js#45770] [Help]
Hello , i am implementing a preloader in Nextjs 13 , for that i used the router.events.on , but the event in not firing  , any ideas ?
Thanks .

```
import {  useEffect } from "react";

import { useRouter } from "next/router";

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  useEffect(() => {
    router.events.on("routeChangeComplete", (url) => {
      console.log(`completely routed to ${url}`);
    });
  }, []);

  return <Component {...pageProps} />;
}
export default MyApp;
```

**Comment by icyJoseph** [comment]
Hi,

It works fine on my end.

What are you expecting? The first load won't fire the event. It is only every navigation after the first load that'll log your message.

The reason why the fist load won't fire event is because the page mounts and loads way before the useEffect happens. Also because there's no route change, only first route loading.

![Screenshot 2023-02-10 at 08 48 43](https://user-images.githubusercontent.com/21013447/218033072-93366b4d-867e-4f54-a675-37c76de2a7be.png)

**Reply by jamiviz** [reply]
Thanks for the explanations , so if i cant use router.events.on for implement a preloader for the first load , what i should use in this case ? useSwr ?

**Reply by icyJoseph** [reply]
At top of your `_app` file you can check if `typeof window !== 'undefined'` and place a window `load` event listener:

```js
window.addEventListener("load", (event) => {
  console.log("page is fully loaded");
 // prefetch some data, or whatever necessary
});
```

So in your case you could do:

```js
import '../styles/globals.css';

import { useEffect } from 'react';

import Router from 'next/router';

if (typeof window !== 'undefined') {
  window.addEventListener('load', (event) => {
    console.log('page is fully loaded');
  });
}

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  useEffect(() => {
    const handler = (url) => {
      console.log(`completely routed to ${url}`);
    };

    Router.events.on('routeChangeComplete', handler);

    return () => Router.events.off('routeChangeComplete', handler);
  }, []);

  return <Component {...pageProps} />;
}

export default MyApp;
```

Do remember that Strict Mode runs components mount+unmount once, so you need to clean up your event handler.

I also use the Router instance, but I guess you could potentially skip that.

- Stackblitz example: https://stackblitz.com/edit/nextjs-c6amvr?file=pages%2F_app.js
   - Open the console (browser) to see the on load event happening, the terminal won't show those!

**Reply by icyJoseph** [reply]
I must say that using link preload (html) is the best way to prefetch data though, but maybe it doesn't accommodate to your use case.

**Comment by daves28** [comment]
Similarly, the `routeChangeCompleted` event handler doesn't fire for me ONLY with minified production code when pressing the browser back button.
